code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/identity/TokenResourceManager.java

```diff
@@ -115,4 +138,50 @@ public void setTokenCreators(List<TokenCreator> tokenCreators) {
         this.tokenCreators = tokenCreators;
     }
 
+    @Override
+    protected Object deleteInternal(String type, String id, Object obj, ApiRequest request) {
+        if (!StringUtils.equals(AbstractTokenUtil.TOKEN, request.getType())) {
+            return null;
+        }
+        return deleteToken(obj, request);
+    }
+
+    protected Object deleteToken(Object obj, ApiRequest request) {
+        String jwt = "";
+        Token token = null;
+        if (SecurityConstants.INTERNAL_AUTH_PROVIDERS.contains(SecurityConstants.AUTH_PROVIDER.get())) {
+            for (TokenCreator tokenCreator : tokenCreators) {
+                if (tokenCreator.isConfigured() && tokenCreator.providerType().equalsIgnoreCase(SecurityConstants.AUTH_PROVIDER.get())) {
+                    if (!SecurityConstants.SECURITY.get()) {
+                        tokenCreator.reset();
+                    }
+                    jwt = tokenCreator.getTokenJWTValue();
+                    token = tokenCreator.getCurrentToken();
+                    break;
+                }
+            }
+        } else {
+            if (externalAuthProvider.isConfigured()) {
+                jwt = tokenUtil.getJWT();
+                token = externalAuthProvider.readCurrentToken();
+            }
+        }
+
```

Comment:
Seems like we can do some optimization between deleteToken and listInternal, since they both call the providers to read the token, this logic can be put in a utility method and called from both deleteToken and listInternal.